oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/ResourceOwnerPasswordCredentialsTokenRequest.java

```diff
@@ -0,0 +1,113 @@
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.auth.oauth2;
+
+import static com.linecorp.armeria.common.auth.oauth2.GrantedOAuth2AccessToken.SCOPE;
+import static java.util.Objects.requireNonNull;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Supplier;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.client.WebClient;
+import com.linecorp.armeria.common.auth.oauth2.ClientAuthorization;
+import com.linecorp.armeria.common.auth.oauth2.GrantedOAuth2AccessToken;
+import com.linecorp.armeria.common.auth.oauth2.InvalidClientException;
+import com.linecorp.armeria.common.auth.oauth2.TokenRequestException;
+import com.linecorp.armeria.common.auth.oauth2.UnsupportedMediaTypeException;
+
+/**
+ * Implements Resource Owner Password Credentials Grant request
+ * as per <a href="https://tools.ietf.org/html/rfc6749#section-4.3">[RFC6749], Section 4.3</a>.
+ */
+class ResourceOwnerPasswordCredentialsTokenRequest extends AbstractAccessTokenRequest {
+
+    private static final String USERNAME = "username";
+    private static final String PASSWORD = "password";
+    private static final String PASSWORD_GRANT_TYPE = PASSWORD;
+
+    private final Supplier<? extends Map.Entry<String, String>> userCredentialsSupplier;
+
+    /**
+     * Implements Resource Owner Password Credentials Grant request/response flow,
+     * as per <a href="https://tools.ietf.org/html/rfc6749#section-4.3">[RFC6749], Section 4.3</a>.
+     *
+     * @param accessTokenEndpoint A {@link WebClient} to facilitate an Access Token request. Must correspond to
+     *                            the Access Token endpoint of the OAuth 2 system.
+     * @param accessTokenEndpointPath A URI path that corresponds to the Access Token endpoint of the
+     *                                OAuth 2 system.
+     * @param clientAuthorization Provides client authorization for the OAuth requests,
+     *                            as per <a href="https://tools.ietf.org/html/rfc6749#section-2.3">[RFC6749], Section 2.3</a>.
+     * @param userCredentialsSupplier A supplier of user credentials: "username" and "password" used to grant
+     *                                the Access Token.
+     */
+    ResourceOwnerPasswordCredentialsTokenRequest(
+            WebClient accessTokenEndpoint, String accessTokenEndpointPath,
+            @Nullable ClientAuthorization clientAuthorization,
+            Supplier<? extends Map.Entry<String, String>> userCredentialsSupplier) {
```

Comment:
Ditto - could use `Supplier<BasicToken>`?

Reply:
Same response as above...
No, it should be very generic. Credentials will come most likely from some kind of "secret" store like Vault (e.g. [LogicalResponse](https://github.com/BetterCloud/vault-java-driver/blob/master/src/main/java/com/bettercloud/vault/response/LogicalResponse.java)).